### DISCUSSION TITLE
Selecting a KDF

### DISCUSSION THREAD

**Discussion by overheadhunter** [encryption-alliance/unified-vault-format#20] [Approved]
### Problem

As currently drafted, we require to use *multiple* keys (possible of varying length) for various algorithms, e.g. 512 bit for file name encryption with SIV and 256 bit for file header encryption with GCM.

Such key material needs to be derived from a *single* vault master key. If this key is sufficiently long, the most simple way would be to split it into specific slices. Experience with other vault format specification tells that this is a rather unflexible solution if future enhancements require further derived keys while avoiding key reuse.

### Proposal

Instead, we will use KDFs to derive subkeys from the master key. To some extent this will even allow stretching a short master key to a longer subkey.

I hereby propose to make use of _One Step KDF_ specified in section 4 of [NIST SP 800-56C REV. 2](https://doi.org/10.6028/NIST.SP.800-56Cr2), using SHA-512 as its hash function.

> [!NOTE]
> The KDF should be specified in `vault.uvf`, allowing future versions of the standard to install a different KDF.

### Rationale

The proposed KDF is easy to implement and thus easy to analyzse, well-known, established and last but not least recommended.

SHA-512 is commonly available on all platforms, including the Web Crypto API, without relying on third party libraries. It is faster than SHA-256 on 64-bit machines and has a potentially larger security strength. The SHA3 family hash functions offer arbitrarily large inputs, but as long as our master key is `< 2^128` bits, this argument is irrelevant. Furthermore resistance against length extension attacks is irrelevant, since such attacks requires knowledge over the master key, making the attack rather pointless in the first place. So there is no real gain in using SHA3, while they aren't that broadly available.

**Comment by overheadhunter** [comment]
After reading section 8.2 in the linked document, I believe it might be beneficial to use `HMAC-SHA512` instead of `SHA-512`:

> In this Recommendation, the MAC algorithms [...] use a salt value as a MAC key [...]. This Recommendation does not require the use of a randomly selected salt value. In particular, [...] this Recommendation specifies that a predetermined default (e.g., all-zero) byte string be used as the salt value.

These are my thoughts:

1. An all-zero (or otherwise fixed) salt means we don't require additional algorithm-specific parameters. Just specifying the KDF in the vault metadata is sufficient.
2. HMAC should be available on all platforms that support the used hash function (or is trivial to implement)
3. While I don't know any specific attack, we should generally presume that attacking any part of a crypto scheme might be easier than bruteforcing the masterkey. And in theory such an attack might involve exploiting a weak hash function, where HMACs may still hold the ground.

**Comment by overheadhunter** [comment]
Alternatively, there are some arguments speaking for `HKDF` (also using `SHA-512` as its hash function):

First and foremost, it is a goal of HKDF "to discourage the proliferation of multiple KDF mechanisms" by proposing a standard that works for most applications, even if the input is not uniformly random.

Furthermore, HKDF *is* a variant of the _Two-Step Key Derivation_ defined in [NIST SP800-56C REV.2](https://doi.org/10.6028/NIST.SP.800-56Cr2):

> [RFC 5869](https://datatracker.ietf.org/doc/html/rfc5869) specifies a version of the above extraction-then-expansion key-derivation procedure using HMAC for both the extraction and expansion steps.

HKDF is commonly available in many environments, including the Web Crypto API.

However, HKDF _should_ be used with a salt, as 

> the use of salt adds significantly to the strength of HKDF

This leads to the requirement to store one additional parameter in `vault.uvf`. However, acting as a _general_ salt this may be multi purpose, e.g. for seeding the encryption of the root directory.

**Reply by tobihagemann** [reply]
HKDF seems to be a good choice. 

I must admit that I'm not sure about a multi-purpose salt though. Should this already be considered here or can we decide on this separately?

**Reply by overheadhunter** [reply]
You're right, let's just have a single-purpose salt for the KDF. Further Salts can easily and deterministically be generated using the KDF itself.

**Comment by rfjakob** [comment]
Hi, I don't think we need to store a salt for the HKDF. Why would we?

To generate several different keys we can put fixed values in the "info" field like done here: https://github.com/rfjakob/gocryptfs/blob/master/internal/cryptocore/hkdf.go

**Reply by rfjakob** [reply]
Reading through the RFC again and also https://crypto.stackexchange.com/questions/97980/hkdf-randomness-extraction-salt-or-no-salt , https://crypto.stackexchange.com/questions/97975/applications-in-which-you-should-shouldnt-use-a-salt-with-hkdf , looks like it's complicated. Showing why it's not needed seems more complicated than having a salt.

I'm retracting the comment, let's have a salt.

**Reply by overheadhunter** [reply]
Making use of `info` is indeed very useful to derive different keys for different parts of the application. Regardless, `info` is independent from the salt, which I agree we don't necessarily "need to store". But it is a recommendation that we can in fact easily fulfil, see PR #21.
